error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java

```diff
@@ -541,6 +541,284 @@ Mono<T> after(Mono<T> mono) {
+        Consumer<D> resourceCleanup,
+        boolean eager) {
+      return Flux.using(resourceSupplier, sourceSupplier, resourceCleanup, eager).single();
+    }
+
+    @AfterTemplate
+    Mono<T> after(
+        Callable<D> resourceSupplier,
+        Function<D, M> sourceSupplier,
+        Consumer<D> resourceCleanup,
+        boolean eager) {
+      return Mono.using(resourceSupplier, sourceSupplier, resourceCleanup, eager);
+    }
+  }
+
+  /**
+   * Don't unnecessarily transform a {@link Flux#usingWhen(Publisher, Function, Function)} to a
+   * mono, instead use the equivalent API provided by {@link Mono}.
+   */
+  static final class MonoUsingWhen<
+      D,
+      T,
+      P extends Publisher<? extends T>,
+      P2 extends Publisher<?>,
+      M extends Mono<? extends T>> {
+    @BeforeTemplate
+    Mono<T> before(
+        Publisher<D> resourceSupplier,
+        Function<D, P> resourceClosure,
+        Function<D, P2> asyncCleanup) {
+      return Flux.usingWhen(resourceSupplier, resourceClosure, asyncCleanup).single();
+    }
+
+    @AfterTemplate
+    Mono<T> after(
+        Publisher<D> resourceSupplier,
+        Function<D, M> resourceClosure,
+        Function<D, P2> asyncCleanup) {
+      return Mono.usingWhen(resourceSupplier, resourceClosure, asyncCleanup);
+    }
+  }
+
+  /**
+   * Don't unnecessarily transform a {@link Flux#usingWhen(Publisher, Function, Function,
+   * BiFunction, Function)} to a mono, instead use the equivalent API provided by {@link Mono}.
+   */
+  static final class MonoUsingWhen2<
+      D,
+      T,
+      P extends Publisher<? extends T>,
+      P2 extends Publisher<?>,
+      M extends Mono<? extends T>> {
+    @BeforeTemplate
+    Mono<T> before(
+        Publisher<D> resourceSupplier,
+        Function<D, P> resourceClosure,
+        Function<D, P2> asyncComplete,
+        BiFunction<D, ? super Throwable, P2> asyncError,
+        Function<D, P2> asyncCancel) {
+      return Flux.usingWhen(
+              resourceSupplier, resourceClosure, asyncComplete, asyncError, asyncCancel)
+          .single();
+    }
+
+    @AfterTemplate
+    Mono<T> after(
+        Publisher<D> resourceSupplier,
+        Function<D, M> resourceClosure,
+        Function<D, P2> asyncComplete,
+        BiFunction<D, ? super Throwable, ? extends Publisher<?>> asyncError,
+        Function<D, P2> asyncCancel) {
+      return Mono.usingWhen(
+          resourceSupplier, resourceClosure, asyncComplete, asyncError, asyncCancel);
+    }
+  }
+
+  /**
+   * Don't unnecessarily transform a {@link Mono#using(Callable, Function)} to a flux, instead use
+   * the equivalent API provided by {@link Flux}.
+   */
+  static final class FluxUsing<
+      D extends AutoCloseable, T, P extends Publisher<? extends T>, M extends Mono<? extends T>> {
+    @BeforeTemplate
+    Flux<T> before(Callable<D> resourceSupplier, Function<D, M> sourceSupplier) {
+      return Mono.using(resourceSupplier, sourceSupplier).flux();
```

Comment:
not sure I agree with this one; using `Mono#flux` at least makes explicit that it can emit at most 1 element (and sometimes a `Flux` is needed to adhere to method signatures)

Reply:
> not sure I agree with this one; using Mono#flux at least makes explicit that it can emit at most 1 element (and sometimes a Flux is needed to adhere to method signatures)

If you check the method signature (Also what I was trying to explain [here](https://github.com/PicnicSupermarket/error-prone-support/pull/1393#discussion_r1827310197)), is that the `Function` of the _sourceSupplier_ explicitly returns a `Mono`, so by definition, we know the whole chain returns one element. So the `Mono` > `Flux` conversion seems redundant to me.

WDYT?

Reply:
`Flux.using` with a `Mono` sourceSupplier looks worse to me as IMO that just obfuscates what's going on, and thus I'd favour the explicit conversion with `flux()`

Reply:
Yeah, that makes sense to me 👌 

Just to make sure I understand your stance, is this a "choosing a lesser evil" type of design choice? Or do you think that what we're attempting to re-write is actually a valid case?

Reply:
> Or do you think that what we're attempting to re-write is actually a valid case

Yeah, I think `Mono#flux` should generally not be rewritten (bar some very specific exceptions where it's ultra-clear the `Flux` variant only supplies 1 element, i.e. with `just` and `error`)

Reply:
Fair argument, I honestly have no strong opinion here.

Let's see what others think, then I will happily close the PR 😄 Thanks @Venorcis

Reply:
I think the other way around is very valid though 😄 i.e. don't use `Flux#single` when you can directly use a `Mono` instead

Reply:
(Though it is hard to know if the pattern was applied because it was intentionally specifying a flux that emits one element; or simply an unawareness of the equivalent APIs in `Flux`; Let's assume the former)

Reply:
> ..using Mono#flux at least makes explicit that it can emit at most 1 element ...

Though it is hard to know if the pattern was applied because it was intentionally specifying a flux that emits one element; or simply an unawareness of the equivalent APIs in `Flux`; Let's assume the former 👍 

Applied ✅